### PR TITLE
Introduce caching to authorized APIs

### PR DESCRIPTION
**Pull request by ThaminduR**
### Proposed changes in this pull request

$subject

### Issue

https://github.com/wso2/product-is/issues/17438

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICache.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    private static final String CACHE_NAME = "AuthorizedAPICache";
+    private static volatile AuthorizedAPICache instance;
+
+    public AuthorizedAPICache(String cacheName) {
+
+        super(cacheName);
+    }
+
+    public static AuthorizedAPICache getInstance() {
```
Comment (SujanSanjula96) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICache.java:36]: add javadoc comments

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICache.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        if (instance == null) {
+            synchronized (AuthorizedAPICache.class) {
+                if (instance == null) {
+                    instance = new AuthorizedAPICache(CACHE_NAME);
+                }
+            }
+        }
+        return instance;
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICache.java:50]: Can't we make this instance a singleton instance by initilizing a priavte static final instance

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedAuthorizedAPIDAOImpl.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+        this.authorizedAPIDAO = authorizedAPIDAO;
+        authorizedAPICache = AuthorizedAPICache.getInstance();
+    }
+
+    @Override
+    public void addAuthorizedAPI(String applicationId, String apiId, String policyId, List<Scope> scopes, int tenantId)
+            throws IdentityApplicationManagementException {
+
+
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedAuthorizedAPIDAOImpl.java:56]:
```suggestion
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedAuthorizedAPIDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        if (authorizedAPI == null) {
+            try {
+                authorizedAPI = authorizedAPIDAO.getAuthorizedAPI(appId, apiId, tenantId);
+                if (authorizedAPI != null) {
+                    addToCache(appId, apiId, authorizedAPI, tenantId);
+                }
+            } catch (IdentityApplicationManagementException e) {
+                String error = "Error while retrieving authorized API for application id: " + appId + " and api id: "
+                        + apiId + " in tenant id: " + tenantId;
+                LOG.error(error, e);
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedAuthorizedAPIDAOImpl.java:104]: no need to log the error here

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedAuthorizedAPIDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                throw new IdentityApplicationManagementException(error, e);
+            }
+        }
+        return authorizedAPI != null ? authorizedAPI : authorizedAPIDAO.getAuthorizedAPI(appId, apiId, tenantId);
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedAuthorizedAPIDAOImpl.java:108]: Do we need to get the resource from DB again if authorizedAPI == null?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICacheEntry.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+package org.wso2.carbon.identity.application.mgt.cache;
+
+import org.wso2.carbon.identity.application.common.model.AuthorizedAPI;
+
+import java.io.Serializable;
+
+/**
+ * Cache entry for authorized API.
+ */
+public class AuthorizedAPICacheEntry implements Serializable {
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICacheEntry.java:28]: Shall we extends CacheEntry instead implements Serializable

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICacheKey.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+package org.wso2.carbon.identity.application.mgt.cache;
+
+import java.io.Serializable;
+
+/**
+ * Cache key for authorized API.
+ */
+public class AuthorizedAPICacheKey implements Serializable {
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/cache/AuthorizedAPICacheKey.java:26]: Shall we extends CacheKey instead implements Serializable

**Review by dewniMW — APPROVED** [review]

**Pull request merged**